internal/parse/parser.go

```diff
@@ -187,57 +215,105 @@ func isNameByte(c byte) bool {
+// construct, if they n't parse they return false, restore the parser and leave
+// the default value in  other return type
 
-		{bytef: p.peekByte, result: []bool{false}, input: "", data: []string{"a"}},
-		{bytef: p.peekByte, result: []bool{false}, input: "b", data: []string{"a"}},
-		{bytef: p.peekByte, result: []bool{true}, input: "a", data: []string{"a"}},
+func (p *Parser) parseIdentifier() (string, bool) {
+	if p.pos >= len(p.input) {
+		return "", false
+	}
+	if p.skipByte('*') {
+		return "*", true
+	}
 
-		{bytef: p.skipByte, result: []bool{false}, input: "", data: []string{"a"}},
-		{bytef: p.skipByte, result: []bool{false}, input: "abc", data: []string{"b"}},
-		{bytef: p.skipByte, result: []bool{true, true}, input: "abc", data: []string{"a", "b"}},
+	idStart := p.pos
+	if p.skipName() {
+		return p.input[idStart:p.pos], true
+	}
+	return "", false
+}
 
-		{bytef: p.skipByteFind, result: []bool{false}, input: "", data: []string{"a"}},
-		{bytef: p.skipByteFind, result: []bool{false, true, true}, input: "abcde", data: []string{"x", "b", "c"}},
-		{bytef: p.skipByteFind, result: []bool{true, false}, input: "abcde ", data: []string{" ", " "}},
+// parseFullName parses a column name or Go field name, optionally dot-prefixed by
+// its table or type name respectively.
+func (p *Parser) parseFullName(idc idClass) (FullName, bool) {
+	cp := p.save()
+	var fn FullName
+	p.skipSpaces()
+	if id, ok := p.parseIdentifier(); ok {
+		fn.Prefix = id
+		if p.skipByte('.') {
+			if id, ok := p.parseIdentifier(); ok {
+				fn.Name = id
+				return fn, true
+			}
+		} else {
+			// A column name specified without a table prefix is a name not a
+			// prefix
+			if idc == columnId {
+				fn.Name = fn.Prefix
+				fn.Prefix = ""
+			}
+			return fn, true
+		}
+	}
+	cp.restore()
+	return fn, false
+}
 
-		{stringf0: p.skipSpaces, result: []bool{false}, input: "", data: []string{}},
-		{stringf0: p.skipSpaces, result: []bool{false}, input: "abc    d", data: []string{}},
-		{stringf0: p.skipSpaces, result: []bool{true}, input: "     abcd", data: []string{}},
-		{stringf0: p.skipSpaces, result: []bool{true}, input: "  \t  abcd", data: []string{}},
-		{stringf0: p.skipSpaces, result: []bool{false}, input: "\t  abcd", data: []string{}},
+// parseInputExpression parses an SDL input go-defined type to be used as a
+// query argument.
+func (p *Parser) parseInputExpression() (*InputPart, bool, error) {
+	cp := p.save()
+	var err error
+	var fn FullName
+	var ok bool
 
-		{stringf: p.skipString, result: []bool{false}, input: "", data: []string{"a"}},
-		{stringf: p.skipString, result: []bool{true, true}, input: "helloworld", data: []string{"hElLo", "w"}},
-		{stringf: p.skipString, result: []bool{true, true}, input: "hello world", data: []string{"hello", " "}},
+	p.skipSpaces()
+	if p.skipByte('$') {
+		fn, ok = p.parseFullName(typeId)
+		if !ok {
+			err = fmt.Errorf("malformed input type")
+		}
+		p.skipSpaces()
+		return &InputPart{fn}, true, err
```

Comment:
This function is returning an error with `ok=true`, which doesn't look right. The `!ok` case above should be immediately bailing to prevent this, and this should return `... , true, nil` instead. 

On a more general note, it's a good practice to return nil/false/zero values in case of errors, except in the rare cases where the intermediate values are actually desired.

Reply:
I've changed the returns so we return nil in all cases except where an error is expected.

I've also added a comment explaining the way the parse functions work with return values. There are three different return states that a parse function can be in as explained in the comment in the code.